### DISCUSSION TITLE
Notification LED animations

### DISCUSSION THREAD

**Discussion by adryzz** [rust3ds/ctru-rs#142] [Ideas]
The notification LED (and more precisely the MCU) allows for animations to be played on the notification LED. (this is the direct way to do this, without patching the `news` service like CTRRGBPAT or similiar homebrew does).

All that is needed is a handle to mcu:HWC to write arbitrary data to an i2c address.

The address is 0x2D and you need to write a 100 byte buffer.

The first 4 bytes are stuff like delay between colors, smoothing, repeat etc, and the remaining 96 bytes is a planar RGB buffer of 32 colors.

It would be nice to provide an interface for this, as it's a very nice piece of hardware just like the QTM, MVD, volume/3D sliders.

i made a quick crate and wrapped `mcuhwc` in `ctru-rs` on another branch, and it all works.

The point of this discussion is to ask whether such infrastructure (the animation code, not the `mcuhwc` service) should be included with `ctru-rs` or be an entirely separate crate.

edit: here's the demo

https://github.com/rust3ds/ctru-rs/assets/46694241/1b2d284b-7e0d-46f4-90f1-0fb3bbdaf70c

**Comment by DeltaF1** [comment]
FWIW there's a [C++ library](https://github.com/PabloMK7/ctr-led-brary) to do this already that uses [this PTM command](https://www.3dbrew.org/wiki/PTMSYSM:SetInfoLEDPattern) rather than accessing i2c directly. (I also ported/translated the library to Rust as a demo of my IPC lib but I never published it cause it was janky).

I'm in favour of exposing the LED stuff in ctru-rs but using the PTM command might be nicer, since I think MCU services are more limited? The wiki says that there can only be one session open per MCU service vs PTM allowing 24 sessions.

**Reply by adryzz** [reply]
as long as you can use the same animation format then yeah
